### PR TITLE
statistic: alter evict policy of resource metering reporter

### PR DESCRIPTION
**Pull request by zhongzc**
### What problem does this PR solve?

Problem Summary:

Previously, the LFU evict policy is based on accumulated CPU time in a minute. In case that the buffer comes full in the first few seconds, the coming records with other tags will be always dropped in the rest of the minute.

### What is changed and how it works?

What's Changed:

Evict policy is based on CPU time in a second. Default `max_resource_groups` is changed to `200`.

### Related changes

- Based on https://github.com/tikv/tikv/pull/10365

### Check List 

Tests 

- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by andylokandy — APPROVED** [review]

**Comment by zhongzc** [comment]
cc @crazycs520 @breeswish

File: components/resource_metering/src/reporter.rs [hunk 1]
```diff
@@ -119,54 +125,84 @@ impl Runnable for ResourceMeteringReporter {
...
-                                cpu_time.push(ms);
-                            }
-                            *total += ms;
-                        }
-                        None => {
-                            self.records
-                                .insert(tag.clone(), (vec![timestamp_secs], vec![ms], ms));
-                        }
-                    }
+                    *tmp_group_map.entry(tag.clone()).or_insert(0) += *ms as u32;
```
Comment (breeswish) [inline components/resource_metering/src/reporter.rs:141]: How about using an Arc to wrap the resource group to reduce memory allocation?

File: components/resource_metering/src/reporter.rs [hunk 2]
```diff
@@ -119,54 +125,84 @@ impl Runnable for ResourceMeteringReporter {
...
+                //   +----------+----------+-------+-------+-------+-------+-------+
+                //   | Threshold           |   30                      ^           |
+                //   +----------+----------+-------+-------+-------+-------+-------+
+                //   | Will be Saved       |   Y   |   Y   |   Y   |   N   |   N   |
+                //   +---------------------+-------+-------+-------+-------+-------+
+                let threshold = (tmp_group_map.len() > max_resource_groups)
+                    .then(|| {
+                        tmp_top_vec.extend(tmp_group_map.values());
+                        let (_, threshold, _) = tmp_top_vec
+                            .select_nth_unstable_by(max_resource_groups, |a, b| b.cmp(a));
```
Comment (breeswish) [inline components/resource_metering/src/reporter.rs:187]: 👍

File: components/resource_metering/src/reporter.rs [hunk 3]
```diff
@@ -119,54 +125,84 @@ impl Runnable for ResourceMeteringReporter {
...
 
-                if self.records.len() > self.config.max_resource_groups {
-                    self.find_top_k.clear();
-                    for (_, _, total) in self.records.values() {
-                        self.find_top_k.push(*total);
+                let timestamp_secs = records.begin_unix_time_secs;
+                let mut other = 0;
+                for (tag, ms) in tmp_group_map.drain() {
+                    if ms > threshold {
+                        let (ts, cpu) = self.records.entry(tag).or_insert((vec![], vec![]));
```
Comment (breeswish) [inline components/resource_metering/src/reporter.rs:196]: Can we use `SmallVec` to reduce allocations?

**Review by breeswish — COMMENTED** [review]
I think there is one additional work:

Previously we keep 200 records for each seconds, and finally just use the records for reporting.

Now we will have more and more records for each seconds (there may be +200 for each second). And just before the reporting we should evict and report only top 200.

File: components/resource_metering/src/reporter.rs (hunk 3 above)
Comment (zhongzc) [reply components/resource_metering/src/reporter.rs:196]: If so, it still needs to allocate later. 

https://github.com/tikv/tikv/blob/3044802f1171ca9edbd16f49214988e3a1a33b3c/components/resource_metering/src/reporter.rs#L241-L249

File: components/resource_metering/src/reporter.rs (hunk 1 above)
Comment (zhongzc) [reply components/resource_metering/src/reporter.rs:141]: It sounds to be complicated. 

It's a good option to use references, but we may either:
1. not to reuse `tmp_group_map` and `tmp_group_vec` or
2. to introduce some `unsafe`s

I prefer `1`: two allocations for eliminating others

**Comment by zhongzc** [comment]
> I think there is one additional work:
> 
> Previously we keep 200 records for each seconds, and finally just use the records for reporting.
> 
> Now we will have more and more records for each seconds (there may be +200 for each second). And just before the reporting we should evict and report only top 200.

Do you mean evict 60+1 times every minute?

**Comment by breeswish** [comment]
> > I think there is one additional work:
> > Previously we keep 200 records for each seconds, and finally just use the records for reporting.
> > Now we will have more and more records for each seconds (there may be +200 for each second). And just before the reporting we should evict and report only top 200.
> 
> Do you mean evict 60+1 times every minute?

@crazycs520 Please take a look about the evict behaviour. What kind of eviction do we have? Please make sure that the TiKV's data evict and TiDB's data evict works identically.

**Pull request closed without merging**